src/main/java/com/gregtechceu/gtceu/api/capability/recipe/ItemRecipeCapability.java

```diff
@@ -339,6 +331,70 @@ public int getMaxParallelRatio(IRecipeCapabilityHolder holder, GTRecipe recipe,
         return minMultiplier;
     }
 
+    private Object2IntMap<ItemStack> getIngredientStacks(IRecipeCapabilityHolder holder) {
+        Object2IntMap<ItemStack> map = new Object2IntOpenHashMap<>();
+        Object2IntMap<ItemStack> result = new Object2IntOpenHashMap<>();
+
+        List<IRecipeHandler<?>> recipeHandlerList = Objects
+                .requireNonNullElseGet(holder.getCapabilitiesProxy().get(IO.IN, ItemRecipeCapability.CAP),
+                        Collections::<IRecipeHandler<?>>emptyList)
+                .stream()
+                .filter(handler -> !handler.isProxy()).toList();
+
+        for (IRecipeHandler<?> container : recipeHandlerList) {
+
+            var itemMap = container.getContents().stream().filter(ItemStack.class::isInstance)
+                    .map(ItemStack.class::cast)
+                    .flatMap(con -> GTHashMaps.fromItemStackCollection(Collections.singleton(con)).object2IntEntrySet()
+                            .stream())
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Integer::sum,
+                            () -> new Object2IntOpenCustomHashMap<>(ItemStackHashStrategy.comparingAllButCount())));
+
+            if (container.isDistinct()) {
+                result.putAll(itemMap);
+            } else {
+                for (Object2IntMap.Entry<ItemStack> obj : itemMap.object2IntEntrySet()) {
+                    if (!isInMap(map, obj.getKey())) {
+                        map.put(obj.getKey(), obj.getIntValue());
+                    } else {
+                        updateMap(map, obj.getKey());
+                    }
+                }
+            }
+        }
+
+        if (!map.isEmpty()) result.putAll(map);
+        return result;
+    }
+
+    private boolean isInMap(Object2IntMap<ItemStack> map, ItemStack is) {
+        for (Object2IntMap.Entry<ItemStack> obj : map.object2IntEntrySet()) {
+            if (obj.getKey().getItem() == is.getItem()) {
+                if (obj.getKey().hasTag() && is.hasTag()) {
+                    return obj.getKey().getTag() == is.getTag();
+                } else {
+                    return true;
+                }
+            }
```

Comment:
This will match a stack with NBT and a stack without one as equal. Use `ItemStackHashStrategy#comparingAllButCount` instead. Same goes for the hashmaps themselves.